zoekt: calculate bucket offsets lazily

stefanhengl opened the pull request:
Instead of creating bucket offsets on read, we calculate the bucket
offset for a given bucket index on the fly.

Why?
Experiments on a larger cluster showed that createBucketOffsets has a
significant impact (>80%) on heap usage. This is still a bit puzzlling, because
I could not reproduce this locally. My intuition tells me this is a bug but while 
I keep debugging we might as well test the approach proposed in this PR.

The potential downside of the lazy buckets is more CPU usage per search.

Test plan:
- All tests, exept tests checking SizeBytes(), pass with `ZOEKT_ENABLE_BTREE=true go test ./...`
- Manually tested search on a corpus of repos
- Inspected heap profiles locally
- Updated units tests

stefanhengl committed:
zoekt: calculate bucket offsets lazily

Instead of creating bucket offsets on read, we calculate the bucket
offset for a given bucket index on the fly.

Experiments on a larger cluster showed that createBucketOffsets has a
significant impact on heap usage. This is a bit puzzlling, because this
does not happen locally.

stefanhengl committed:
typo

keegancsmith commented on read.go:
```diff
@@ -534,31 +534,14 @@ func (d *indexData) newBtreeIndex(toc *indexTOC) (btreeIndex, error) {
-func createBucketOffsets(sec simpleSection, bucketSize int) []uint32 {
-	step := uint32((bucketSize / 2) * ngramEncoding)
-
-	offsets := make([]uint32, 0, ((sec.sz-1)/step)+1)
```
sz - 1 will overflow for empty contents, sz is unsigned, so I reckon this is the root cause of the bug.

keegancsmith committed:
store lastBucketIndex to try make invariants clearer in getBucket

Also moved btree specific construction code into a freeze method. Also
added a testcase where the final bucket size is less than bucketSize.

Co-authored-by: Stefan Hengl <stefan@sourcegraph.com>

stefanhengl committed:
getBucket doesn't need bucketSize after all
